File: tools/update_readme.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Gera/atualiza automaticamente o README.md principal do projeto
a partir dos README.md dos diretórios dia_XX.

Regras adotadas:
- # Dia XX        -> ignorado (já representado no README principal)
- ## Título       -> item numerado
- ### Subtítulo   -> subitem indentado
- Âncoras seguem o padrão aproximado do GitHub (PT-BR friendly)
"""

import re
import unicodedata
from pathlib import Path


ROOT = Path(".")
OUT = ROOT / "README.md"


def gh_slug(text: str) -> str:
    """
    Gera âncora compatível com o GitHub:
    - lowercase
    - mantém acentos
    - remove pontuação
    - espaços viram hífens
    """
    text = text.strip().lower()

    # remove pontuação, mas mantém letras acentuadas
    text = re.sub(r"[^\w\s\-áéíóúàèìòùâêîôûãõç]", "", text, flags=re.UNICODE)

    # espaços -> hífen
    text = re.sub(r"\s+", "-", text)

    # colapsa múltiplos hífens
    text = re.sub(r"-{2,}", "-", text).strip("-")

    return text


def read_headings(md_path: Path):
    """
    Lê headings Markdown (##, ###, etc.), ignorando blocos de código.
    Retorna lista de tuplas: (nível, título).
    """
    headings = []
    in_code_block = False

    for line in md_path.read_text(encoding="utf-8").splitlines():
        if line.strip().startswith("```"):
            in_code_block = not in_code_block
            continue

        if in_code_block:
            continue

        match = re.match(r"^(#{1,6})\s+(.+?)\s*$", line)
        if not match:
            continue

        level = len(match.group(1))
        title = match.group(2).strip()
        headings.append((level, title))

    return headings


def day_label_and_anchor(day_dir_name: str):
    """
    Converte:
      dia_02 -> ("DIA 02", "dia-02")
    """
    number = day_dir_name.split("_")[1]
    label = f"DIA {number}"
    anchor = gh_slug(label)
    return label, anchor


def main():
    day_dirs = sorted(
        [
            p for p in ROOT.iterdir()
            if p.is_dir() and re.match(r"^dia_\d{2}$", p.name)
        ],
        key=lambda p: p.name,
    )

    lines = []
    lines.append("# Workshop Go: do Zero à API")
    lines.append("")

    global_index = 1

    for day_dir in day_dirs:
        day_label, day_anchor = day_label_and_anchor(day_dir.name)
        day_link = f"../../tree/main/{day_dir.name}#{day_anchor}"

        lines.append(f"## [{day_label}]({day_link})")
        lines.append("")

        day_readme = day_dir / "README.md"
        if not day_readme.exists():
            lines.append("_README.md não encontrado para este dia._")
            lines.append("")
            continue

        headings = read_headings(day_readme)

        has_current_item = False
        sub_index = 0

        for level, title in headings:
            # Ignora "# Dia XX"
            if level == 1:
                continue

            # Item principal
            if level == 2:
                anchor = gh_slug(title)
                link = f"../../tree/main/{day_dir.name}#{anchor}"
                lines.append(f"{global_index}. [{title}]({link})")
                lines.append("")
                global_index += 1
                sub_index = 0
                has_current_item = True
                continue

            # Subitem
            if level == 3 and has_current_item:
                sub_index += 1
                anchor = gh_slug(title)
                link = f"../../tree/main/{day_dir.name}#{anchor}"
                lines.append(f"    {sub_index}. [{title}]({link})")
                continue

        lines.append("")

    OUT.write_text("\n".join(lines).rstrip() + "\n", encoding="utf-8")
    print("README.md atualizado com sucesso.")


if __name__ == "__main__":
    main()

File: tools/update_readme.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode"
)

var (
	reDayDir  = regexp.MustCompile(`^dia_\d{2}$`)
	reHeading = regexp.MustCompile(`^(#{1,6})\s+(.+?)\s*$`)
	reSpace   = regexp.MustCompile(`\s+`)
	reHyphens = regexp.MustCompile(`-{2,}`)

	projectOut = "README.md"
)

type Heading struct {
	Level int
	Title string
}

// slugger implementa o comportamento do GitHub para headings duplicados:
// primeiro uso: "titulo"
// duplicados: "titulo-1", "titulo-2", ...
type slugger struct {
	seen map[string]int
}

func newSlugger() *slugger {
	return &slugger{seen: make(map[string]int)}
}

func (s *slugger) Anchor(title string) string {
	base := ghSlugBase(title)
	if base == "" {
		base = "section"
	}
	// GitHub: primeiro uso sem sufixo; a partir do segundo, -1, -2, ...
	if n, ok := s.seen[base]; ok {
		n++
		s.seen[base] = n
		return fmt.Sprintf("%s-%d", base, n)
	}
	s.seen[base] = 0
	return base
}

func main() {
	root := "."
	dayDirs, err := listDayDirs(root)
	if err != nil {
		fatalf("falha ao listar diretórios dia_XX: %v", err)
	}

	var lines []string
	lines = append(lines, "# Workshop Go: do Zero à API", "")

	globalIdx := 1

	for _, dayDir := range dayDirs {
		dayLabel := dayLabelFromDir(dayDir)
		// Âncora do H1 "# Dia XX" dentro do README do dia
		dayAnchor := ghSlugBase(dayLabel)
		dayLink := fmt.Sprintf("../../tree/main/%s#%s", dayDir, dayAnchor)

		lines = append(lines, fmt.Sprintf("## [%s](%s)", dayLabel, dayLink), "")

		dayReadme := filepath.Join(root, dayDir, "README.md")
		if _, err := os.Stat(dayReadme); err != nil {
			if os.IsNotExist(err) {
				lines = append(lines, "_README.md não encontrado para este dia._", "")
				continue
			}
			fatalf("erro ao acessar %s: %v", dayReadme, err)
		}

		headings, err := readHeadings(dayReadme)
		if err != nil {
			fatalf("falha ao ler headings de %s: %v", dayReadme, err)
		}

		// Slugger por arquivo (igual ao GitHub)
		slugs := newSlugger()

		hasCurrentItem := false
		subIdx := 0

		for _, h := range headings {
			switch h.Level {
			case 1:
				// Ainda assim, registramos o H1 no slugger para garantir
				// consistência caso alguém crie links para ele.
				_ = slugs.Anchor(h.Title)
				continue

			case 2:
				anchor := slugs.Anchor(h.Title)
				link := fmt.Sprintf("../../tree/main/%s#%s", dayDir, anchor)
				lines = append(lines, fmt.Sprintf("%d. [%s](%s)", globalIdx, h.Title, link), "")
				globalIdx++
				subIdx = 0
				hasCurrentItem = true

			case 3:
				if !hasCurrentItem {
					// Ainda assim registra para manter contagem correta em duplicados
					_ = slugs.Anchor(h.Title)
					continue
				}
				subIdx++
				anchor := slugs.Anchor(h.Title)
				link := fmt.Sprintf("../../tree/main/%s#%s", dayDir, anchor)
				lines = append(lines, fmt.Sprintf("    %d. [%s](%s)", subIdx, h.Title, link))

			default:
				// Opcional: registrar para duplicados em níveis mais profundos
				_ = slugs.Anchor(h.Title)
				continue
			}
		}

		lines = append(lines, "")
	}

	out := strings.TrimRight(strings.Join(lines, "\n"), "\n") + "\n"
	if err := os.WriteFile(projectOut, []byte(out), 0o644); err != nil {
		fatalf("falha ao escrever %s: %v", projectOut, err)
	}

	fmt.Println("README.md atualizado com sucesso.")
}

func listDayDirs(root string) ([]string, error) {
	entries, err := os.ReadDir(root)
	if err != nil {
		return nil, err
	}

	var dayDirs []string
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		if reDayDir.MatchString(name) {
			dayDirs = append(dayDirs, name)
		}
	}

	sort.Strings(dayDirs)
	return dayDirs, nil
}

func readHeadings(mdPath string) ([]Heading, error) {
	f, err := os.Open(mdPath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var headings []Heading
	sc := bufio.NewScanner(f)

	inCodeBlock := false

	for sc.Scan() {
		line := sc.Text()
		trim := strings.TrimSpace(line)

		// Toggle bloco de código quando linha começa com ```
		if strings.HasPrefix(trim, "```") {
			inCodeBlock = !inCodeBlock
			continue
		}
		if inCodeBlock {
			continue
		}

		m := reHeading.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		level := len(m[1])
		title := strings.TrimSpace(m[2])
		headings = append(headings, Heading{Level: level, Title: title})
	}

	if err := sc.Err(); err != nil {
		return nil, err
	}
	return headings, nil
}

func dayLabelFromDir(dayDir string) string {
	// dia_02 -> "DIA 02"
	parts := strings.Split(dayDir, "_")
	n := ""
	if len(parts) == 2 {
		n = parts[1]
	}
	return "DIA " + n
}

func ghSlugBase(title string) string {
	// GitHub-like base:
	// - lowercase
	// - mantém acentos
	// - remove pontuação/símbolos
	// - espaços -> hífen
	// - colapsa hífens
	s := strings.ToLower(strings.TrimSpace(title))

	// Mantém: letras (incluindo acentuadas), dígitos, underscore, espaço, hífen
	var buf bytes.Buffer
	for _, r := range s {
		switch {
		case unicode.IsLetter(r) || unicode.IsDigit(r):
			buf.WriteRune(r)
		case r == '_' || r == ' ' || r == '-':
			buf.WriteRune(r)
		default:
			// descarta pontuação/símbolos
		}
	}

	s = buf.String()
	s = reSpace.ReplaceAllString(s, "-")
	s = reHyphens.ReplaceAllString(s, "-")
	s = strings.Trim(s, "-")
	return s
}

func fatalf(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}
